File: src/leetcode/trees/rust/src/_543.rs
use crate::eval::{constructor, Tree, TreeNumTestCase};
use crate::utils::parse_json_array;

fn recursive_max_depth_binary_tree(root: Option<Tree>) -> i64 {
    root.iter()
        .map(|r| {
            recursive_max_depth_binary_tree(r.borrow().left.clone())
                .max(recursive_max_depth_binary_tree(r.borrow().right.clone()))
                + 1
        })
        .sum()
}

fn diameter(root: Option<Tree>) -> i64 {
    match root {
        Some(node) => {
            let node_ref = node.borrow();
            let root_diameter = recursive_max_depth_binary_tree(node_ref.left.clone())
                + recursive_max_depth_binary_tree(node_ref.right.clone());
            let left_diameter = diameter(node_ref.left.clone());
            let right_diameter = diameter(node_ref.right.clone());
            root_diameter.max(left_diameter).max(right_diameter)
        }
        None => 0,
    }
}

pub fn main(test_cases: Vec<TreeNumTestCase>) {
    for test_case in test_cases {
        let input: Vec<Option<i64>> = parse_json_array::<i64>(&test_case.input);
        let expected_output: i64 = test_case.output;

        let input_tree = constructor(&input);
        let output: i64 = diameter(input_tree);

        assert_eq!(output, expected_output, "Test Case {:?} failed", input);
    }
    println!("Tests ran successfully for problem ID 543");
}

File: src/leetcode/trees/cpp/543.cpp
#include "eval.cpp"
#include "stddef.h"
#include <algorithm>
#include <cstdio>
#include <sys/param.h>
#include <vector>

using namespace std;

int recursive_max_depth_binary_tree(Node *root) {
  if (root != nullptr) {
    int leftDepth = recursive_max_depth_binary_tree(root->left);
    int rightDepth = recursive_max_depth_binary_tree(root->right);

    return MAX(leftDepth, rightDepth) + 1;
  }

  return 0;
}

int diameter(Node *root) {
  if (root != nullptr) {
    int root_diameter = recursive_max_depth_binary_tree(root->left) +
                        recursive_max_depth_binary_tree(root->right);
    int left_diameter = diameter(root->left);
    int right_diameter = diameter(root->right);

    return max({root_diameter, left_diameter, right_diameter});
  }

  return 0;
}

int main(int argc, char *argv[]) {
  // Parse arguments
  vector<int> input = parse_single_array(argv[1]);
  int expected_output = atoi(argv[2]);

  // Create Trees and get depth
  Node *inputTree = constructor(input);
  int output = diameter(inputTree);

  if (output != expected_output) {
    printf("Error getting diameter (543) for provided input tree expected %d "
           "got %d\n",
           expected_output, output);
    return 1;
  }

  return 0;
}
